File: packages/sdk/src/commands.ts
/**
 * `sandbox.commands` - run shell commands inside a sandbox.
 *
 * Hits the per-sandbox data plane with the access token. A paused sandbox
 * (401 stale token, or 503 because the VM isn't running) is transparently
 * resumed via `POST /activate` and the call is retried once — see
 * `tokenRetry.ts`.
 *
 * Accessed as `sandbox.commands.run(...)`.
 */

import { spawnCommand } from "./commandSession.js"
import { dataPlaneTarget } from "./config.js"
import { SandboxError } from "./errors.js"
import { request, streamSSE } from "./http.js"
import { withTokenRetry } from "./tokenRetry.js"
import type {
  ApiExecResult,
  ApiExecStreamEvent,
  CommandOptions,
  CommandResult,
  CommandSession,
  SpawnOptions,
} from "./types.js"

/** @internal */
export interface CommandsDeps {
  sandboxId: string
  sandboxHost: string
  getAccessToken: () => string
  refreshActivate: () => Promise<string>
}

/**
 * Max bytes read from a non-streaming `/exec` response before the SDK aborts
 * the read and throws `ValidationError`.
 *
 * The data plane is untrusted: `boxd`'s sync exec handler buffers the command's
 * full stdout/stderr in the VM and returns it in one JSON body, and neither the
 * data-plane proxy nor `fetch` caps that body — so a command can stream back as
 * much output as the sandbox has RAM. Buffering it unbounded into memory is a
 * DoS vector for any shared host (e.g. the hosted MCP server). Callers that
 * genuinely need large output should stream via `onStdout`/`onStderr` instead.
 */
export const MAX_EXEC_RESPONSE_BYTES = 10 * 1024 * 1024 // 10 MiB

export class Commands {
  private readonly _dataPlaneBaseUrl: string
  private readonly _routingHeaders: Record<string, string>

  /** @internal */
  constructor(private readonly _deps: CommandsDeps) {
    const target = dataPlaneTarget(_deps.sandboxId, _deps.sandboxHost)
    this._dataPlaneBaseUrl = target.url
    this._routingHeaders = target.headers
  }

  /**
   * Execute a command inside the sandbox.
   *
   * If `onStdout` or `onStderr` callbacks are provided, the command is
   * streamed via SSE. Otherwise, it runs synchronously (waits for completion).
   *
   * Paused sandboxes are transparently resumed before execution.
   *
   * @example
   * ```typescript
   * // Synchronous
   * const result = await sandbox.commands.run("echo hello")
   * console.log(result.stdout) // "hello\n"
   *
   * // Streaming
   * const result = await sandbox.commands.run("npm start", {
   *   onStdout: (data) => process.stdout.write(data),
   *   onStderr: (data) => process.stderr.write(data),
   *   timeoutMs: 120_000,
   * })
   * ```
   */
  async run(
    command: string,
    options: CommandOptions = {},
  ): Promise<CommandResult> {
    const { cwd, env, timeoutMs, onStdout, onStderr } = options
    const isStreaming = onStdout !== undefined || onStderr !== undefined

    const body: Record<string, unknown> = { command }
    if (cwd !== undefined) body.working_dir = cwd
    if (env !== undefined) body.env = env
    // API expects seconds; convert from ms
    if (timeoutMs !== undefined) body.timeout_s = Math.ceil(timeoutMs / 1000)

    if (isStreaming) {
      return this._runStreaming(body, options)
    }
    return this._runSync(body, options)
  }

  /**
   * Spawn a command and return a live, full-duplex session.
   *
   * Unlike `run` (one-shot), `spawn` hands back a handle while the process is
   * still running: stream output via `onStdout`/`onStderr`, write to its
   * `stdin`, `kill` it, and `await session.wait()` for the exit result.
   *
   * Release the session when you're done (use `await using`, or call `wait()`
   * or `close()`) so the socket and its reader are cleaned up.
   *
   * A paused sandbox is resumed transparently before the session opens.
   *
   * @example
   * ```typescript
   * const session = await sandbox.commands.spawn("python -i", {
   *   onStdout: (data) => process.stdout.write(data),
   * })
   * session.stdin.write("print(2 + 2)\n")
   * session.stdin.close()
   * const { exitCode } = await session.wait()
   * ```
   */
  spawn(command: string, options: SpawnOptions = {}): Promise<CommandSession> {
    return spawnCommand(this._deps, command, options)
  }

  private async _runSync(
    body: Record<string, unknown>,
    options: CommandOptions,
  ): Promise<CommandResult> {
    const send = (token: string) =>
      request<ApiExecResult>({
        method: "POST",
        url: `${this._dataPlaneBaseUrl}/exec`,
        headers: { ...this._routingHeaders, "X-Access-Token": token },
        body,
        // Add a 5s buffer so the server-side command timeout fires first
        // and returns its proper response before the client aborts.
        timeoutMs:
          options.timeoutMs !== undefined
            ? options.timeoutMs + 5_000
            : undefined,
        signal: options.signal,
        // The data plane is untrusted — bound the buffered response so a
        // command's runaway stdout/stderr can't exhaust the caller's memory.
        maxBytes: options.maxOutputBytes ?? MAX_EXEC_RESPONSE_BYTES,
      })

    const raw = await withTokenRetry(this._deps, send)
    return {
      stdout: raw.stdout ?? "",
      stderr: raw.stderr ?? "",
      exitCode: raw.exit_code ?? 0,
    }
  }

  private async _runStreaming(
    body: Record<string, unknown>,
    options: CommandOptions,
  ): Promise<CommandResult> {
    const send = async (token: string) =>
      this._consumeStream(
        `${this._dataPlaneBaseUrl}/exec/stream`,
        { ...this._routingHeaders, "X-Access-Token": token },
        body,
        options,
      )
    // Safe for streaming: the resumable status (401/503) is returned before any
    // SSE data is written, so a retry can't double-emit callbacks.
    return withTokenRetry(this._deps, send)
  }

  private async _consumeStream(
    url: string,
    headers: Record<string, string>,
    body: Record<string, unknown>,
    options: CommandOptions,
  ): Promise<CommandResult> {
    let stdout = ""
    let stderr = ""
    let exitCode = 0
    let error: string | undefined
    let sawFinished = false

    await streamSSE({
      url,
      headers,
      body,
      timeoutMs:
        options.timeoutMs !== undefined ? options.timeoutMs + 5_000 : undefined,
      signal: options.signal,
      onEvent: (event: ApiExecStreamEvent) => {
        if (event.stdout) {
          stdout += event.stdout
          options.onStdout?.(event.stdout)
        }
        if (event.stderr) {
          stderr += event.stderr
          options.onStderr?.(event.stderr)
        }
        if (event.finished) {
          sawFinished = true
          exitCode = event.exit_code ?? 0
          error = event.error
        }
      },
    })

    if (!sawFinished) {
      throw new SandboxError(
        "Command stream ended without a finished event (possible network disconnect)",
      )
    }

    if (error) {
      stderr += error
    }

    return { stdout, stderr, exitCode }
  }
}

File: packages/ruby-sdk/lib/superserve/commands.rb
module Superserve
  # `sandbox.commands` — run shell commands inside a sandbox.
  #
  # Two modes:
  # - Synchronous: waits for the command to finish, returns stdout/stderr/exit_code.
  # - Streaming: when on_stdout / on_stderr callbacks are passed, fires them via
  #   SSE and returns the aggregated result at the end.
  #
  # Paused sandboxes are transparently resumed by the platform before execution.
  class Commands
    # @api private
    def initialize(base_url, sandbox_id, api_key, connection: nil)
      @base_url = base_url
      @sandbox_id = sandbox_id
      @api_key = api_key
      @connection = connection
    end

    def run(command, cwd: nil, env: nil, timeout_seconds: nil, on_stdout: nil, on_stderr: nil)
      body = { command: command }
      body[:working_dir] = cwd if cwd
      body[:env] = env if env
      body[:timeout_s] = timeout_seconds if timeout_seconds

      headers = { "X-API-Key" => @api_key }
      streaming = on_stdout || on_stderr

      if streaming
        run_streaming(body, headers, on_stdout, on_stderr, timeout_seconds)
      else
        run_sync(body, headers, timeout_seconds)
      end
    end

    private

    def run_sync(body, headers, timeout_seconds)
      effective_timeout = timeout_seconds ? timeout_seconds + 5 : HTTP::DEFAULT_TIMEOUT
      raw = HTTP.api_request(
        :post,
        "#{@base_url}/sandboxes/#{@sandbox_id}/exec",
        headers: headers,
        json_body: body,
        timeout: effective_timeout,
        connection: @connection
      )
      CommandResult.new(
        stdout: raw["stdout"] || "",
        stderr: raw["stderr"] || "",
        exit_code: raw["exit_code"] || 0
      )
    end

    def run_streaming(body, headers, on_stdout, on_stderr, timeout_seconds)
      stdout_parts = []
      stderr_parts = []
      exit_code = 0
      saw_finished = false
      effective_timeout = timeout_seconds ? timeout_seconds + 5 : HTTP::DEFAULT_STREAM_TIMEOUT

      HTTP.stream_sse(
        "#{@base_url}/sandboxes/#{@sandbox_id}/exec/stream",
        headers: headers,
        json_body: body,
        timeout: effective_timeout,
        connection: @connection
      ) do |event|
        if event["stdout"]
          stdout_parts << event["stdout"]
          on_stdout&.call(event["stdout"])
        end
        if event["stderr"]
          stderr_parts << event["stderr"]
          on_stderr&.call(event["stderr"])
        end
        if event["finished"]
          saw_finished = true
          exit_code = event["exit_code"] || 0
          stderr_parts << event["error"] if event["error"]
        end
      end

      unless saw_finished
        raise SandboxError.new(
          "Command stream ended without a finished event (possible network disconnect)"
        )
      end

      CommandResult.new(
        stdout: stdout_parts.join,
        stderr: stderr_parts.join,
        exit_code: exit_code
      )
    end
  end
end
